bin/changelog/zenhub/requests.js

```diff
@@ -0,0 +1,101 @@
+'use strict';
+
+/* eslint no-console: 0 */
+
+const ZenHub = require( 'zenhub-api' );
+const { REPO } = require( '../config' );
+const { authedGraphql } = require( '../common' );
+const { pull } = require( 'lodash' );
+
+const api = new ZenHub( process.env.ZH_API_KEY );
+
+const getQuery = ( after ) => {
+	const [ owner, repo ] = REPO.split( '/' );
+	const paging = after ? `, after: "${ after }"` : '';
+	const query = `
+	{
+		repository(owner: "${ owner }", name: "${ repo }") {
+			pullRequests(first: 100, states: [MERGED], orderBy: {field: CREATED_AT, direction: ASC}${ paging }) {
+				totalCount
+				pageInfo {
+					hasNextPage
+					endCursor
+				}
+				nodes {
+					number
+					title
+					url
+					author {
+						login
+					}
+					body
+					labels(last: 10) {
+						nodes {
+							name
+						}
+					}
+				}
+			}
+		}
+	}
+	`;
+	return query;
+};
+
+const fetchAllIssuesForRelease = async ( releaseId ) => {
+	const releaseIssues = await api.getReleaseReportIssues( {
+		release_id: releaseId,
+	} );
+	return releaseIssues.map( ( releaseIssue ) => releaseIssue.issue_number );
+};
+
+const extractPullRequestsMatchingReleaseIssue = (
+	releaseIds,
+	pullRequests
+) => {
+	return pullRequests.filter( ( pullRequest ) => {
+		const hasPullRequest = releaseIds.includes( pullRequest.number );
+		if ( hasPullRequest ) {
+			pull( releaseIds, pullRequest.number );
+			return true;
+		}
+		return false;
+	} );
+};
+
+const fetchAllPullRequests = async ( releaseId ) => {
+	// first get all release issue ids
+	const releaseIds = await fetchAllIssuesForRelease( releaseId );
+	let maxPages = Math.ceil( releaseIds.length / 100 ) + 2;
+	const fetchResults = async ( after ) => {
```

Comment:
The zenhub script doesn’t seem to work, it never returns a changelog even though both zenhub and gihtub return results, I'm looking into it

Reply:
it seems the problem is that by doing the changes I made in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1543/commits/ecef6e810d2dae0605d52932683ef12d8e4c4d33 by using `first` and `after`, the zenhub script, which was brute-forcing through all the PRs, will start from the oldest PR ever, meaning it will never reach the PR of the latest release, in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1543/commits/060985d05a7c2918dc26518c10ce2b48ce9cad64 I reverted to the original approach, but it still won’t work for releases with over 100 PR, since after the first hit, it will return 0-99 then 1-100 then 2-101, meaning in total 102 results, not 300 hits as we would assume, I would leave it like this until we can figure something out later

Reply:
keeping the same logic and using API V3 (REST API) should solve this issue for us, we gain nothing significant from graphql expect shaping our data differently, we're still limited by 100 results per hit

Reply:
> we gain nothing significant from graphql expect shaping our data differently, 

I wouldn't be so quick to jump to this conclusion. Using v3, resulted in two extra requests:

- [one for the collaborator](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/bin/changelog.js#L30-L34)
- [one for whether the pull is merged](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/master/bin/changelog.js#L30-L34)

With graphql we drop that to one extra request (which can save over 100 extra requests per changelog generation).  Unless maybe you discovered a way to get the merge info in the v3 response?

I'm all for going back to API v3 if we have to, but it seems odd we aren't able to get this working with API v4. I'd like to keep trying myself if that's okay?

Reply:
> I wouldn't be so quick to jump to this conclusion. Using v3, resulted in two extra requests

I did miss that, I think we can drop the collaborator hit for now and think about what we can check if the PR is merged by checking the value of `merged_at` if it’s null or not, it will be `null` for unmerged ones, and a valid Date for merged ones.

Yes, I still feel like V4 pagination should work, I didn’t give it much time so another look from you would be great